xds/src/main/java/io/grpc/xds/LookasideLb.java

```diff
@@ -49,169 +48,223 @@
+import io.grpc.xds.XdsSubchannelPickers.ErrorPicker;
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.logging.Logger;
 import javax.annotation.Nullable;
 
-/** Lookaside load balancer that handles balancer name changes. */
-final class LookasideLb extends ForwardingLoadBalancer {
+/** Lookaside load balancer that handles EDS config. */
+final class LookasideLb extends LoadBalancer {
 
-  private final EdsUpdateCallback edsUpdateCallback;
-  private final GracefulSwitchLoadBalancer lookasideChannelLb;
+  private final ChannelLogger channelLogger;
+  private final EndpointUpdateCallback endpointUpdateCallback;
+  private final GracefulSwitchLoadBalancer switchingLoadBalancer;
   private final LoadBalancerRegistry lbRegistry;
   private final LocalityStoreFactory localityStoreFactory;
   private final LoadReportClientFactory loadReportClientFactory;
-
-  private String balancerName;
-
-  LookasideLb(Helper lookasideLbHelper, EdsUpdateCallback edsUpdateCallback) {
+  private final Bootstrapper bootstrapper;
+  private final Helper lookasideLbHelper;
+
+  // Most recent XdsConfig.
+  // Becomes non-null once handleResolvedAddresses() successfully.
+  @Nullable
+  private XdsConfig xdsConfig;
+  // Most recent EndpointWatcher.
+  // Becomes non-null once handleResolvedAddresses() successfully.
+  @Nullable
+  private EndpointWatcher endpointWatcher;
+
+  // Becomes non-null and calls getObject() once handleResolvedAddresses() successfully.
+  // Will call returnObject() at balancer shutdown.
+  @Nullable
+  private ObjectPool<XdsClient> xdsClientRef;
+  // Becomes non-null once handleResolvedAddresses() successfully.
+  @Nullable
+  XdsClient xdsClient;
+  // Becomes non-null for EDS-only case once handleResolvedAddresses() successfully.
+  // TODO(zdapeng): Stop using it once XdsClientImpl is used.
+  @Nullable
+  ManagedChannel channel;
+
+  LookasideLb(Helper lookasideLbHelper, EndpointUpdateCallback endpointUpdateCallback) {
     this(
         lookasideLbHelper,
-        edsUpdateCallback,
+        endpointUpdateCallback,
         LoadBalancerRegistry.getDefaultRegistry(),
         LocalityStoreFactory.getInstance(),
-        LoadReportClientFactory.getInstance());
+        LoadReportClientFactory.getInstance(),
+        Bootstrapper.getInstance());
   }
 
   @VisibleForTesting
   LookasideLb(
       Helper lookasideLbHelper,
-      EdsUpdateCallback edsUpdateCallback,
+      EndpointUpdateCallback endpointUpdateCallback,
       LoadBalancerRegistry lbRegistry,
       LocalityStoreFactory localityStoreFactory,
-      LoadReportClientFactory loadReportClientFactory) {
-    this.edsUpdateCallback = edsUpdateCallback;
+      LoadReportClientFactory loadReportClientFactory,
+      Bootstrapper bootstrapper) {
+    this.lookasideLbHelper = lookasideLbHelper;
+    this.channelLogger = lookasideLbHelper.getChannelLogger();
+    this.endpointUpdateCallback = endpointUpdateCallback;
```

Comment:
nit: `this.endpointUpdateCallback` and `this.lookasideLbHelper` are being used elsewhere without any kind of null checks here.

Reply:
Done.